pkg/sdk/sql_builder.go

```diff
@@ -0,0 +1,251 @@
+			// if there is not, then we need to recurse into the struct
+			// and get the clauses from there
+			ddlTag := field.Tag.Get("ddl")
+			if ddlTag != "" {
+				ddlTagParts := strings.Split(ddlTag, ",")
+				if ddlTagParts[0] == "keyword" {
+					clauses = append(clauses, sqlClauseKeyword{
+						value: field.Tag.Get("db"),
+						qt:    getQuoteTypeFromTag(field.Tag, "ddl"),
+					})
+				}
+			}
+			innerClauses, err := b.parseStruct(value.Interface())
+			if err != nil {
+				return nil, err
+			}
+			clauses = append(clauses, innerClauses...)
+			continue
+		}
+
+		// default case, if not a struct then it is a field
+		clause := b.parseField(field, value)
+		if clause != nil {
+			clauses = append(clauses, clause)
+		}
+	}
+	return clauses, nil
+}
+
+// parseField parses an exported struct field and returns a sqlClause.
+func (b *sqlBuilder) parseField(field reflect.StructField, value reflect.Value) sqlClause {
+	if field.Tag.Get("ddl") == "" {
+		return nil
+	}
+
+	ddlTag := strings.Split(field.Tag.Get("ddl"), ",")[0]
+	dbTag := field.Tag.Get("db")
+
+	switch ddlTag {
+	case "static":
+		return sqlClauseStatic(dbTag)
+	case "keyword":
+		if value.Kind() == reflect.Bool {
+			useKeyword := value.Interface().(bool)
+			if useKeyword {
+				return sqlClauseKeyword{
+					value: dbTag,
+					qt:    getQuoteTypeFromTag(field.Tag, "ddl"),
+				}
+			}
+			return nil
+		}
+		return sqlClauseKeyword{
+			value: value.Interface().(string),
+			qt:    getQuoteTypeFromTag(field.Tag, "ddl"),
+		}
+	case "command":
+		return sqlClauseCommand{
+			key:   dbTag,
+			value: value.Interface().(string),
+			qt:    getQuoteTypeFromTag(field.Tag, "ddl"),
+		}
+	case "parameter":
+		return sqlClauseParameter{
+			key:   dbTag,
+			value: value.Interface(),
+			qt:    getQuoteTypeFromTag(field.Tag, "ddl"),
+		}
+	default:
+		return nil
+	}
+}
+
+// parseUnexportedField parses an unexported struct field and returns a sqlClause.
+func (b *sqlBuilder) parseUnexportedField(field reflect.StructField, value reflect.Value) sqlClause {
+	if field.Tag.Get("ddl") == "" {
+		return nil
+	}
+	tagParts := strings.Split(field.Tag.Get("ddl"), ",")
+	ddlType := tagParts[0]
+	dbTag := field.Tag.Get("db")
+	switch ddlType {
+	case "static":
+		return sqlClauseStatic(dbTag)
+	case "name":
+		return sqlClauseName(value.String())
+	case "keyword":
+		if value.Kind() == reflect.Bool {
+			useKeyword := value.Bool()
+			if !useKeyword {
+				return nil
+			}
+		}
+		return sqlClauseKeyword{
+			value: value.String(),
+			qt:    getQuoteTypeFromTag(field.Tag, "ddl"),
+		}
+	}
+
+	return nil
+}
+
+type sqlClause interface {
+	String() string
+}
+
+type sqlClauseStatic string
+
+func (v sqlClauseStatic) String() string {
+	return string(v)
+}
+
+type sqlClauseKeyword struct {
+	value string
+	qt    quoteType
+}
+
+func (v sqlClauseKeyword) String() string {
+	// make sure we dont double quote the keyword
+	trimmedValue := strings.Trim(v.value, v.qt.String())
```

Comment:
trim only removes the characters from the beginning or the end of the string, but not from the middle, eg:
- Trim("abaca", "a") == "bac"
- Trim(`"aa"aa"`, `"`) == `aa"aa`
So it's not a solution here.

According to [the docs](https://docs.snowflake.com/en/sql-reference/identifiers-syntax), double quote characters should be doubled when the string is double-quoted so we should do something like this:
```suggestion
	if (v.qt == DoubleQuotes) {
		escapedValue := strings.ReplaceAll(v.value, v.qt.String(), v.qt.String()+v.qt.String())
	} else {
		escapedValue := v.value
	}
```

Reply:
Maybe add a `Quote(s string) string` method to `quoteType` to factorise the escaping logic? That way we can do:
```go
func (v sqlClauseKeyword) String() string {
	return v.qt.Quote(v.value)
}
```
and reuse in other places.